### DISCUSSION TITLE
variable in other scene is returning nil even when set

### DISCUSSION THREAD

**Discussion by TrainzMarcel** [gilzoide/godot-lua-pluginscript#46] [Q&A] [answered]
so i have two scenes;
"basevehicle" and "suspensionarm" (child of spatial named "wheels" which is a child of basevehicle)

in suspensionarm.lua, im only putting data (variables)
and in basevehicle.lua im putting all the logic

under basevehicle i put a node named "wheels" to contain the suspensionarms

![image](https://user-images.githubusercontent.com/108669283/224307116-0737fa40-8998-4a93-83c5-f6afd88bcba8.png)

when i try to assign a node to v.ray (declared as suspensionarm.ray = nil in its script or not) it still prints nil, why is this happening?

**Comment by gilzoide** [comment, marked as answer]
Hey @TrainzMarcel.
Well, setting something to `nil` in Lua is the same as deleting it from the table. So to declare an Object property with default value of `null`, you should use `suspensionarm.ray = Object.null`. This would declare a property in the script's manifest, so it could be accessed by GDScript or by Godot Objects in Lua via `get/get_indexed/set/set_indexed`.

Now, if you had the Lua script instance that is attached to the Object instead of the Object itself, it wouldn't matter if the property was declared or not to Godot and you would be able to set/get it without any fuzz, Lua table style. You can get the Lua script instance today by using `GD.get_lua_instance(v)`. But now that I'm thinking of it, in Lua we probably always want to have the Lua script instance and not the script owner, which is a raw Godot Object pointer 🤔

Ok, so for now, either declare the property using `Object.null` instead of `nil`, or use `GD.get_lua_instance(some_object)` to work with the Lua script instance instead of the Object pointer.
But in a future release (possibly the next one), I guess we should always unbox the Lua script instance from Object Variants automatically. Sounds like a much better developer experience to me ^^

Thanks for the insight!

**Reply by TrainzMarcel** [reply]
thanks!! im happy i can continue working on my data driven approach

**Reply by TrainzMarcel** [reply]
also, i think i would prefer if u shifted ur focus to making lua work in gdextension, since godot 4.0 is the hot new thing (no pressure)

**Reply by gilzoide** [reply]
> also, i think i would prefer if u shifted ur focus to making lua work in gdextension, since godot 4.0 is the hot new thing (no pressure)

Well, I guess you're right. Thanks for the encouragement ^^

**Comment by gilzoide** [comment]
Oh, also, next time you could paste your code using code blocks instead of posting an image. This is useful so that people could copy/paste the code to test it out.

You use triple backticks (\`\`\`) + language name to start a syntax-colored code block (like: \`\`\`lua). Then another triple backticks to end the block.

It appears like this:
```lua
-- This will appear as lua code
local msg = "Hello Lua"
print(msg)
```

**Reply by TrainzMarcel** [reply]
will do

**Comment by gilzoide** [comment]
FYI, I created issue #47 regarding automatic unboxing of Lua script instances.
